### PR TITLE
Add readme guidelines for writing tests

### PR DESCRIPTION
**Pull request by jcantrill**
I'm trying to capture the guidelines and assumptions from #443 and https://trello.com/c/UBias2xZ for writing tests using the framework provided.

### CODE REVIEW

File: test/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Guidelines for Writing Tests
+
+## Assumptions & Requirements
+
+1. Directory paths should be well formed such that they can be run from any directory.
+1. Test scripts in the `$HOME/test` directory assume they are run against a functional Openshift cluster that was setup prior to the execution of the test script.
```
Comment (stevekuznetsov) [inline test/README.md]: `$HOME` here could be misleading

File: test/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+1. Each script in `$HOME/test` should has a companion script in `$HOME/hack/testing`
```
Comment (stevekuznetsov) [inline test/README.md]: ... which can be used to launch the test and do requisite setup steps...

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+1. Each script in `$HOME/test` performs cleanup logic to reset the state of the cluster and logging stack
+
+```
+#!/bin/bash
+
+source "$(dirname "${BASH_SOURCE[0]}" )/../hack/lib/init.sh"
+
+os::util::environment::use_sudo
```
Comment (stevekuznetsov) [inline test/README.md]: let people figure out they need this, don't suggest it outright

File: test/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Guidelines for Writing Tests
+
+## Assumptions & Requirements
+
+1. Directory paths should be well formed such that they can be run from any directory.
+1. Test scripts in the `$HOME/test` directory assume they are run against a functional Openshift cluster that was setup prior to the execution of the test script.
+1. Each script in `$HOME/test` should has a companion script in `$HOME/hack/testing`
+1. Each script in `$HOME/test` performs cleanup logic to reset the state of the cluster and logging stack
+
+```
```
Comment (stevekuznetsov) [inline test/README.md]: use 

    ```sh

for highlighting

File: test/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#!/bin/bash
+
+source "$(dirname "${BASH_SOURCE[0]}" )/../hack/lib/init.sh"
+
+os::util::environment::use_sudo
+
+os::test::junit::declare_suite_start "test/<TESTNAME>"
```
Comment (stevekuznetsov) [inline test/README.md]: declare end

File: test/README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+if [ -n "${DEBUG:-}" ] ; then
```
Comment (stevekuznetsov) [inline test/README.md]: don't suggest this -- `os::cmd` + `set -x` = thousands of lines of output nobody needs

Comment (stevekuznetsov) [inline test/README.md]: there is a `$OS_DEBUG` var that turns on printing from `os::log::debug` functions, use instead

File: test/README.md (hunk 5 above)
Comment (jcantrill) [inline test/README.md]: What do you mean as the change here?

Comment (stevekuznetsov) [inline test/README.md]: If you declare start, declare end too in this example

File: test/README.md [hunk 7]
```diff
@@ -3,23 +3,18 @@
 ## Assumptions & Requirements
 
 1. Directory paths should be well formed such that they can be run from any directory.
-1. Test scripts in the `$HOME/test` directory assume they are run against a functional Openshift cluster that was setup prior to the execution of the test script.
-1. Each script in `$HOME/test` should has a companion script in `$HOME/hack/testing`
-1. Each script in `$HOME/test` performs cleanup logic to reset the state of the cluster and logging stack
+1. Test scripts in the `origin-aggregated-logging/test` directory assume they are run against a functional Openshift cluster that was setup prior to the execution of the test script.
+1. Each script in `origin-aggregated-logging/test` should has a companion script in `origin-aggregated-logging/hack/testing` which can be used to launch the test and do requisite setup steps.
```
Comment (richm) [inline test/README.md]: I don't think this is necessary.  I thought the `/hack/testing` scripts were only to allow us to run the `/test` scripts using `logging.sh`.

File: test/README.md (hunk 5 above)
Comment (richm) [inline test/README.md]: In practice, `declare_suite_end` will be called from `os::test::junit::reconcile_output` which should be the last thing in the `cleanup` function.

**Comment by richm** [comment]
What's the rationale for calling `os::cleanup::all` from an individual test?  e.g. in `test/curator.sh` do we need to `dump_events`, `dump_etcd`, `dump_container_logs`, `dump_pprof_output`, etc.?  Or is that something we would generally do after `entrypoint.sh` is completed? or perhaps only upon failure of an individual test?

**Comment by stevekuznetsov** [comment]
@richm +1 to cleanup in the entrypoint. The tests themselves should only clean up if they make disruptive changes to the cluster

File: test/README.md (hunk 7 above)
Comment (jcantrill) [inline test/README.md]: I defer to you and @stevekuznetsov regarding this statement.  I thought it was need for any custom setup to run the script in /test.  If that is not the case then I will remove it; you tell me.

File: test/README.md [hunk 8]
```diff
@@ -21,7 +21,7 @@ os::test::junit::declare_suite_start "test/<TESTNAME>"
 
 ## Role of Companion scripts in hack/testing
 
-The wrapper or companion script for each test is responsible for 
+The wrapper or companion script for each test is responsible for
```
Comment (richm) [inline test/README.md]: @stevekuznetsov is this true?  I thought the purpose of the `hack/testing/test-*` scripts was to bridge us until we could get rid of `hack/testing`

Comment (stevekuznetsov) [reply test/README.md]: If the only setup, etc that is necessary lives in `entrypoint.sh` then yes, these are not necessary. I think some of these scripts right now call into the test 2x, one for ops and one for not

Comment (stevekuznetsov) [reply test/README.md]: That's kind of a question of how you want to structure your test cases TBH and I don't really have a strong opinion -- but it does seem cleaner to say "we expect these components to work this way" then call that test twice with two different sets of components

Comment (richm) [reply test/README.md]: > I think some of these scripts right now call into the test 2x, one for ops and one for not

I don't think so.  Some of the tests, like curator.sh, will run with and without ops _inside_ the test script itself.

Comment (stevekuznetsov) [reply test/README.md]: Yeah, looks like drift has happened -- whatever you feel is appropriate here should be enforced so it's a little more clear

**Pull request merged**